

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>msprime.intervals &mdash; msprime 0.1.dev1 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/jupyter-sphinx.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/thebelab.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
        <script src="../../_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.0/dist/embed-amd.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home" alt="Documentation Home"> msprime
          

          
          </a>

          
            
            
              <div class="version">
                0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../what_is.html">What is msprime?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
</ul>
<p class="caption"><span class="caption-text">Running simulations</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../ancestry.html">Ancestry</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../mutations.html">Mutations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../demography.html">Demography</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../likelihoods.html">Likelihoods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../utilities.html">Utilities</a></li>
</ul>
<p class="caption"><span class="caption-text">API documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../quickref.html">API quick reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cli.html">Command line interface</a></li>
</ul>
<p class="caption"><span class="caption-text">Developers</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../development.html">Developer documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../CITATION.html">Citing msprime</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html">PORTING IN PROGRESS: Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">PORTING IN PROGRESS: API</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">msprime</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>msprime.intervals</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for msprime.intervals</h1><div class="highlight"><pre>
<span></span><span class="c1">#</span>
<span class="c1"># Copyright (C) 2020 University of Oxford</span>
<span class="c1">#</span>
<span class="c1"># This file is part of msprime.</span>
<span class="c1">#</span>
<span class="c1"># msprime is free software: you can redistribute it and/or modify</span>
<span class="c1"># it under the terms of the GNU General Public License as published by</span>
<span class="c1"># the Free Software Foundation, either version 3 of the License, or</span>
<span class="c1"># (at your option) any later version.</span>
<span class="c1">#</span>
<span class="c1"># msprime is distributed in the hope that it will be useful,</span>
<span class="c1"># but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="c1"># MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="c1"># GNU General Public License for more details.</span>
<span class="c1">#</span>
<span class="c1"># You should have received a copy of the GNU General Public License</span>
<span class="c1"># along with msprime.  If not, see &lt;http://www.gnu.org/licenses/&gt;.</span>
<span class="c1">#</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Utilities for working with intervals and interval maps.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>


<span class="c1"># TODO this is a minimal implementation of the functionality we need to get</span>
<span class="c1"># varying-rates-of-things-along the genome into msprime. It&#39;s deliberately</span>
<span class="c1"># lightweight, and unencumbered by any backing classes from the low-level</span>
<span class="c1"># model. The intention would be to add a bunch of useful functionality here</span>
<span class="c1"># using numpy APIs.</span>
<div class="viewcode-block" id="RateMap"><a class="viewcode-back" href="../../utilities.html#msprime.RateMap">[docs]</a><span class="k">class</span> <span class="nc">RateMap</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A class mapping a numeric value to a set of adjacent intervals along the</span>
<span class="sd">    genome.</span>

<span class="sd">    :param list position: A list of :math:`n+1` positions, starting at 0, and ending</span>
<span class="sd">        in the sequence length over which the RateMap will apply.</span>
<span class="sd">    :param list rate: A list of :math:`n` positive rates that apply between each</span>
<span class="sd">        position.</span>
<span class="sd">    :param float start_position: When returning a mean_rate over the genome, ignore the</span>
<span class="sd">        part of the map from zero up to ``start_position``, as long as the rate is 0 in</span>
<span class="sd">        that region. If ``None`` (default), the ``start_position`` is 0.</span>
<span class="sd">    :param float end_position: When returning a mean_rate over the genome, ignore the</span>
<span class="sd">        part of the map from ``end_position`` onwards, as long as the rate is 0 in that</span>
<span class="sd">        region. If ``None`` (default), the ``end_position`` is the end of the map.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">position</span><span class="p">,</span>
        <span class="n">rate</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">start_position</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">end_position</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="c1"># Making the arrays read-only guarantees rate and cumulative mass stay in sync</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_position</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">position</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_position</span><span class="o">.</span><span class="n">flags</span><span class="o">.</span><span class="n">writeable</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_position</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">size</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Must have at least two positions&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_position</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;First position must be zero&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_position</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">):</span>
            <span class="n">bad_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_position</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Position values not strictly increasing at indexes </span><span class="si">{</span><span class="n">bad_pos</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_rate</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">rate</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_rate</span><span class="o">.</span><span class="n">flags</span><span class="o">.</span><span class="n">writeable</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_rate</span><span class="p">)</span> <span class="o">!=</span> <span class="n">size</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Rate array must have one less entry than the position array&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_rate</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">):</span>
            <span class="n">bad_rates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_rate</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Rate values negative at indexes </span><span class="si">{</span><span class="n">bad_rates</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_mass</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_position</span><span class="p">)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">_rate</span><span class="p">),</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_mass</span><span class="o">.</span><span class="n">flags</span><span class="o">.</span><span class="n">writeable</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="k">if</span> <span class="n">start_position</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">start_position</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_cumulative_mass</span><span class="p">(</span><span class="n">start_position</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Rates before the start_position must all be zero&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_start_position</span> <span class="o">=</span> <span class="n">start_position</span>

        <span class="k">if</span> <span class="n">end_position</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">end_position</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_cumulative_mass</span><span class="p">(</span><span class="n">end_position</span><span class="p">)</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_mass</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Rates after the end_position must all be zero&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_end_position</span> <span class="o">=</span> <span class="n">end_position</span>

        <span class="c1"># Make all of the internal arrays read-only, so they can&#39;t get out of sync</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">position</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The positions between which a constant rate applies. There will be</span>
<span class="sd">        one more position than the number of rates, with the first position being 0</span>
<span class="sd">        and the last being the :attr:`sequence_length`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_position</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">rate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The array of constant rates that apply between each :attr:`position`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_rate</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">cumulative_mass</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The integral of the rates along the genome, one value for each</span>
<span class="sd">        :attr:`position`. The first value will be 0 and the last will be the</span>
<span class="sd">        :attr:`total_mass`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_mass</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">start_position</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The physical start of the map. This position can be greater than 0 for maps</span>
<span class="sd">        which have no recorded rate at the start of a chromosome (often the telomeric</span>
<span class="sd">        region). It primarily affects the :attr:`mean_rate` returned for this rate map.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_start_position</span>

    <span class="nd">@start_position</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">start_position</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">val</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
            <span class="s2">&quot;Cannot change a RateMap start_position: use ratemap.slice(... trim=False)&quot;</span>
        <span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">end_position</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The physical end of the map. This position can be less than the</span>
<span class="sd">        :attr:`sequence_length` for maps which have no recorded rate at the end of a</span>
<span class="sd">        chromosome (often the telomeric region). It primarily affects the</span>
<span class="sd">        :attr:`mean_rate` returned for this rate map.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_end_position</span>

    <span class="nd">@end_position</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">end_position</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">val</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
            <span class="s2">&quot;Cannot change a RateMap end_position: use ratemap.slice(... trim=False)&quot;</span>
        <span class="p">)</span>

<div class="viewcode-block" id="RateMap.uniform"><a class="viewcode-back" href="../../utilities.html#msprime.RateMap.uniform">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">uniform</span><span class="p">(</span><span class="n">sequence_length</span><span class="p">,</span> <span class="n">rate</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a uniform rate map</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">RateMap</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">],</span> <span class="p">[</span><span class="n">rate</span><span class="p">])</span></div>

    <span class="k">def</span> <span class="nf">asdict</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;position&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="s2">&quot;rate&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">rate</span><span class="p">}</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">sequence_length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The sequence length covered by this map</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">total_mass</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The cumulative total over the entire map</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Since we insist that the rate up to start_position and past end_position is 0</span>
        <span class="c1"># we can just return the cumulative total.</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">cumulative_mass</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">mean_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The mean rate over this map, from :attr:`start_position` to</span>
<span class="sd">        :attr:`end_position`, weighted by the span covered by each rate.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">span</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">end_position</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">start_position</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_mass</span> <span class="o">/</span> <span class="n">span</span>

<div class="viewcode-block" id="RateMap.get_rate"><a class="viewcode-back" href="../../utilities.html#msprime.RateMap.get_rate">[docs]</a>    <span class="k">def</span> <span class="nf">get_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the rate at a list of user-specified positions along the map, i.e.</span>
<span class="sd">        interpolated between the fixed :attr:`position` values. Any positions</span>
<span class="sd">        beyond the end  of the map will return the final rate value.</span>

<span class="sd">        :param numpy.ndarray x: The positions for which to return values.</span>

<span class="sd">        :return: An array of rates, the same length as ``x``.</span>
<span class="sd">        :rtype: numpy.ndarray(dtype=np.float64)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">loc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">loc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">minimum</span><span class="p">(</span><span class="n">loc</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rate</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>  <span class="c1"># Return final rate if off the end</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">loc</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No rate exists for positions below the minimum map pos&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">rate</span><span class="p">[</span><span class="n">loc</span><span class="p">]</span></div>

<div class="viewcode-block" id="RateMap.get_cumulative_mass"><a class="viewcode-back" href="../../utilities.html#msprime.RateMap.get_cumulative_mass">[docs]</a>    <span class="k">def</span> <span class="nf">get_cumulative_mass</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the cumulative rate at a list of user-specified positions along the map,</span>
<span class="sd">        i.e. interpolated between the fixed :attr:`position` values.</span>

<span class="sd">        :param numpy.ndarray x: The positions for which to return values.</span>

<span class="sd">        :return: An array of cumulative rates, the same length as ``x``</span>
<span class="sd">        :rtype: numpy.ndarray(dtype=np.float64)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Cannot have physical positions &lt; 0 or &gt; </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">interp</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">cumulative_mass</span><span class="p">)</span></div>

<div class="viewcode-block" id="RateMap.slice"><a class="viewcode-back" href="../../utilities.html#msprime.RateMap.slice">[docs]</a>    <span class="k">def</span> <span class="nf">slice</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">trim</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>  <span class="c1"># noqa:A003</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a subset of this rate map between the specified end points.</span>

<span class="sd">        :param float start: The start of the region to keep. If ``None``, it defaults to</span>
<span class="sd">            0.</span>
<span class="sd">        :param float end: The end of the region to keep. If end is ``None``, it defaults</span>
<span class="sd">            to the end of the map.</span>
<span class="sd">        :param bool trim: If True, remove the flanking regions such that the</span>
<span class="sd">            sequence length of the new rate map is ``end`` - ``start``. If ``False``</span>
<span class="sd">            (default), do not change the coordinate system, but instead replace each</span>
<span class="sd">            flanking region with 0, and set the ``start_position`` and ``end_position``</span>
<span class="sd">            of the returned map to ``start`` and ``end``, inserting these positions if</span>
<span class="sd">            they do not exist.</span>

<span class="sd">        :return: A new RateMap instance</span>
<span class="sd">        :rtype: RateMap</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">start</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">start</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">end</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">end</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span>
            <span class="n">j</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>

        <span class="k">if</span> <span class="p">(</span>
            <span class="n">start</span> <span class="o">&lt;</span> <span class="mi">0</span>
            <span class="ow">or</span> <span class="n">end</span> <span class="o">&lt;</span> <span class="mi">0</span>
            <span class="ow">or</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span>
            <span class="ow">or</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span>
            <span class="ow">or</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="n">end</span>
        <span class="p">):</span>
            <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid subset: start=</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="s2">, end=</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">start</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">i</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
                <span class="n">i</span> <span class="o">-=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">end</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
            <span class="n">j</span> <span class="o">=</span> <span class="n">i</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="n">i</span><span class="p">:],</span> <span class="n">end</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="n">j</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]:</span>
                <span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="n">position</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">rate</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rate</span><span class="p">[</span><span class="n">i</span> <span class="p">:</span> <span class="n">j</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">position</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">start</span>
        <span class="n">position</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">end</span>

        <span class="k">if</span> <span class="n">trim</span><span class="p">:</span>
            <span class="c1"># Return trimmed map with changed coords</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span><span class="n">position</span> <span class="o">-</span> <span class="n">start</span><span class="p">,</span> <span class="n">rate</span><span class="p">)</span>

        <span class="c1"># Need to check regions before &amp; after sliced region are filled out:</span>
        <span class="k">if</span> <span class="n">start</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">rate</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">position</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># Extend</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">rate</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">rate</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>  <span class="c1"># Prepend</span>
                <span class="n">position</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">position</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">end</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">rate</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">position</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span>  <span class="c1"># Extend</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">rate</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rate</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>  <span class="c1"># Append</span>
                <span class="n">position</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">position</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">(</span><span class="n">position</span><span class="p">,</span> <span class="n">rate</span><span class="p">,</span> <span class="n">start_position</span><span class="o">=</span><span class="n">start</span><span class="p">,</span> <span class="n">end_position</span><span class="o">=</span><span class="n">end</span><span class="p">)</span></div>

    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Use slice syntax for obtaining a rate map subset. E.g.</span>
<span class="sd">            &gt;&gt;&gt; rate_map_4m_to_5m = rate_map[4e6:5e6]</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="nb">slice</span><span class="p">)</span> <span class="ow">or</span> <span class="n">key</span><span class="o">.</span><span class="n">step</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;Only interval slicing is supported&quot;</span><span class="p">)</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">key</span><span class="o">.</span><span class="n">stop</span>
        <span class="k">if</span> <span class="n">start</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span>
        <span class="k">if</span> <span class="n">end</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">end</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">end</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">slice</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="n">end</span><span class="p">,</span> <span class="n">trim</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<div class="viewcode-block" id="RateMap.read_hapmap"><a class="viewcode-back" href="../../utilities.html#msprime.RateMap.read_hapmap">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">read_hapmap</span><span class="p">(</span>
        <span class="n">fileobj</span><span class="p">,</span>
        <span class="n">sequence_length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="o">*</span><span class="p">,</span>
        <span class="n">has_header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">position_col</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">rate_col</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">map_col</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
        <span class="c1"># Black barfs with an INTERNAL_ERROR trying to reformat this docstring,</span>
        <span class="c1"># so we explicitly disable reformatting here.</span>
        <span class="c1"># fmt: off</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parses the specified file in HapMap format and returns a RateMap.</span>
<span class="sd">        HapMap files must white-space-delimited, and by default are assumed to</span>
<span class="sd">        contain a single header line (which is ignored). Each subsequent line</span>
<span class="sd">        then contains a physical position (in base pairs) and either a genetic</span>
<span class="sd">        map position (in centiMorgans) or a recombination rate (in centiMorgans</span>
<span class="sd">        per megabase); the rate between the current physical position</span>
<span class="sd">        (inclusive) and the physical position on the next line (exclusive) is</span>
<span class="sd">        taken as constant. By default, the second column of the file is taken</span>
<span class="sd">        as the physical position and the fourth column is taken as the genetic</span>
<span class="sd">        position, as seen in the following sample of the format::</span>

<span class="sd">            Chromosome	Position(bp)  Rate(cM/Mb)  Map(cM)</span>
<span class="sd">            chr10       48232         0.1614       0.002664</span>
<span class="sd">            chr10       48486         0.1589       0.002705</span>
<span class="sd">            chr10       50009         0.159        0.002947</span>
<span class="sd">            chr10       52147         0.1574       0.003287</span>
<span class="sd">            ...</span>
<span class="sd">            chr10	133762002     3.358        181.129345</span>
<span class="sd">            chr10	133766368     0.000        181.144008</span>

<span class="sd">        .. note::</span>
<span class="sd">            The rows are all assumed to come from the same contig, and the</span>
<span class="sd">            first column is currently ignored. Therefore if you have a single</span>
<span class="sd">            file containing several contigs or chromosomes, you must must split</span>
<span class="sd">            it up into multiple files, and pass each one separately to this</span>
<span class="sd">            function.</span>

<span class="sd">        :param str fileobj: Filename or file to read. This is passed directly</span>
<span class="sd">            to :func:`numpy.loadtxt`, so if the filename extension is .gz or .bz2,</span>
<span class="sd">            the file is decompressed first</span>
<span class="sd">        :param float sequence_length: The total length of the map. If ``None``,</span>
<span class="sd">            then assume it is the last physical position listed in the file.</span>
<span class="sd">            Otherwise it must be greater then or equal to the last physical</span>
<span class="sd">            position in the file, and the region between the last physical position</span>
<span class="sd">            and the sequence_length is padded with a rate of zero.</span>
<span class="sd">        :param bool has_header: If True (default), assume the file has a header row</span>
<span class="sd">            and ignore the first line of the file.</span>
<span class="sd">        :param int position_col: The zero-based index of the column in the file</span>
<span class="sd">            specifying the physical position in base pairs. If ``None`` (default)</span>
<span class="sd">            assume an index of 1 (i.e. the second column).</span>
<span class="sd">        :param int rate_col: The zero-based index of the column in the file</span>
<span class="sd">            specifying the rate in cM/Mb. If ``None`` (default) do not use the rate</span>
<span class="sd">            column, but calculate rates using the genetic map positions, as</span>
<span class="sd">            specified in ``map_col``. If the rate column is used, the</span>
<span class="sd">            :attr:`RateMap.start_position` of the returned map is set to the first</span>
<span class="sd">            physical position in the file, and the last value in the rate column</span>
<span class="sd">            must be zero.</span>
<span class="sd">        :param int map_col: The zero-based index of the column in the file</span>
<span class="sd">            specifying the genetic map position in centiMorgans. If ``None``</span>
<span class="sd">            (default), assume an index of 3 (i.e. the fourth column). If the first</span>
<span class="sd">            genetic position is 0, set the :attr:`RateMap.start_position` of the</span>
<span class="sd">            returned map to the first physical position in the file. Otherwise, act</span>
<span class="sd">            as if an additional row, specifying physical position 0 and genetic</span>
<span class="sd">            position 0, exists at the start of the file.</span>
<span class="sd">        :return: A RateMap object.</span>
<span class="sd">        :rtype: RateMap</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># fmt: on</span>
        <span class="n">column_defs</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># column definitions passed to np.loadtxt</span>
        <span class="k">if</span> <span class="n">rate_col</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">map_col</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Default to map_col</span>
            <span class="n">map_col</span> <span class="o">=</span> <span class="mi">3</span>
        <span class="k">elif</span> <span class="n">rate_col</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">map_col</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Cannot specify both rate_col and map_col&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">map_col</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">column_defs</span><span class="p">[</span><span class="n">map_col</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;map&quot;</span><span class="p">,</span> <span class="nb">float</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">column_defs</span><span class="p">[</span><span class="n">rate_col</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;rate&quot;</span><span class="p">,</span> <span class="nb">float</span><span class="p">)</span>
        <span class="n">position_col</span> <span class="o">=</span> <span class="mi">1</span> <span class="k">if</span> <span class="n">position_col</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">position_col</span>
        <span class="k">if</span> <span class="n">position_col</span> <span class="ow">in</span> <span class="n">column_defs</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Cannot specify the same columns for position_col and &quot;</span>
                <span class="s2">&quot;rate_col or map_col&quot;</span>
            <span class="p">)</span>
        <span class="n">column_defs</span><span class="p">[</span><span class="n">position_col</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span>

        <span class="n">column_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">column_defs</span><span class="o">.</span><span class="n">values</span><span class="p">()]</span>
        <span class="n">column_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">loadtxt</span><span class="p">(</span>
            <span class="n">fileobj</span><span class="p">,</span>
            <span class="n">skiprows</span><span class="o">=</span><span class="mi">1</span> <span class="k">if</span> <span class="n">has_header</span> <span class="k">else</span> <span class="mi">0</span><span class="p">,</span>
            <span class="n">dtype</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">column_defs</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span>
            <span class="n">usecols</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">column_defs</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
            <span class="n">unpack</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">column_names</span><span class="p">,</span> <span class="n">column_data</span><span class="p">))</span>

        <span class="k">if</span> <span class="s2">&quot;map&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
            <span class="k">assert</span> <span class="s2">&quot;rate&quot;</span> <span class="ow">in</span> <span class="n">data</span>
            <span class="k">if</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;rate&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The last entry in the &#39;rate&#39; column must be zero&quot;</span><span class="p">)</span>
            <span class="n">pos_Mb</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="mf">1e6</span>
            <span class="n">map_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;rate&quot;</span><span class="p">][:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">pos_Mb</span><span class="p">))</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;map&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">map_pos</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">/</span> <span class="mi">100</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">data</span><span class="p">[</span><span class="s2">&quot;map&quot;</span><span class="p">]</span> <span class="o">/=</span> <span class="mi">100</span>  <span class="c1"># Convert centiMorgans to Morgans</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;map&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Empty hapmap file&quot;</span><span class="p">)</span>

        <span class="c1"># TO DO: read in chrom name from col 0 and poss set as .name</span>
        <span class="c1"># attribute on the RateMap</span>

        <span class="n">physical_positions</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">]</span>
        <span class="n">genetic_positions</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;map&quot;</span><span class="p">]</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">physical_positions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">physical_positions</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">genetic_positions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">start</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;The map distance at the start of the chromosome must be zero&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">physical_positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">physical_positions</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">genetic_positions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># Exception for a map that starts &gt; 0cM: include the start rate</span>
                <span class="c1"># in the mean</span>
                <span class="n">start</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">genetic_positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">genetic_positions</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">sequence_length</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">sequence_length</span> <span class="o">&lt;</span> <span class="n">end</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;The sequence_length cannot be less that the last physical position &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; (</span><span class="si">{</span><span class="n">physical_positions</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="n">sequence_length</span> <span class="o">&gt;</span> <span class="n">end</span><span class="p">:</span>
                <span class="n">physical_positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">physical_positions</span><span class="p">,</span> <span class="n">sequence_length</span><span class="p">)</span>
                <span class="n">genetic_positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">genetic_positions</span><span class="p">,</span> <span class="n">genetic_positions</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

        <span class="k">assert</span> <span class="n">genetic_positions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
        <span class="k">return</span> <span class="n">RateMap</span><span class="p">(</span>
            <span class="n">physical_positions</span><span class="p">,</span>
            <span class="n">rate</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">genetic_positions</span><span class="p">)</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">physical_positions</span><span class="p">),</span>
            <span class="n">start_position</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
            <span class="n">end_position</span><span class="o">=</span><span class="n">end</span><span class="p">,</span>
        <span class="p">)</span></div></div>


<span class="c1"># The RecombinationMap class is deprecated since 1.0. We maintain the</span>
<span class="c1"># functionality where it is possible to do so.</span>

<span class="c1"># TODO update the documentation to make it clear that this is a legacy</span>
<span class="c1"># interface and is deprecated.</span>


<div class="viewcode-block" id="RecombinationMap"><a class="viewcode-back" href="../../utilities.html#msprime.RecombinationMap">[docs]</a><span class="k">class</span> <span class="nc">RecombinationMap</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A RecombinationMap represents the changing rates of recombination</span>
<span class="sd">    along a chromosome. This is defined via two lists of numbers:</span>
<span class="sd">    ``positions`` and ``rates``, which must be of the same length.</span>
<span class="sd">    Given an index j in these lists, the rate of recombination</span>
<span class="sd">    per base per generation is ``rates[j]`` over the interval</span>
<span class="sd">    ``positions[j]`` to ``positions[j + 1]``. Consequently, the first</span>
<span class="sd">    position must be zero, and by convention the last rate value</span>
<span class="sd">    is also required to be zero (although it is not used).</span>

<span class="sd">    .. warning::</span>
<span class="sd">        The ``num_loci`` parameter is deprecated.</span>

<span class="sd">    :param list positions: The positions (in bases) denoting the</span>
<span class="sd">        distinct intervals where recombination rates change. These can</span>
<span class="sd">        be floating point values.</span>
<span class="sd">    :param list rates: The list of rates corresponding to the supplied</span>
<span class="sd">        ``positions``. Recombination rates are specified per base,</span>
<span class="sd">        per generation.</span>
<span class="sd">    :param int num_loci: **This parameter is deprecated**.</span>
<span class="sd">        The maximum number of non-recombining loci</span>
<span class="sd">        in the underlying simulation. By default this is set to</span>
<span class="sd">        the largest possible value, allowing the maximum resolution</span>
<span class="sd">        in the recombination process. However, for a finite sites</span>
<span class="sd">        model this can be set to smaller values.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">positions</span><span class="p">,</span> <span class="n">rates</span><span class="p">,</span> <span class="n">num_loci</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">map_start</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">num_loci</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">num_loci</span> <span class="o">==</span> <span class="n">positions</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;num_loci is no longer supported and should not be used.&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;num_loci does not match sequence length. &quot;</span>
                    <span class="s2">&quot;To set a discrete number of recombination sites, &quot;</span>
                    <span class="s2">&quot;scale positions to span the desired number of loci &quot;</span>
                    <span class="s2">&quot;and set discrete=True&quot;</span>
                <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">map</span> <span class="o">=</span> <span class="n">RateMap</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">rates</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">map_start</span> <span class="o">=</span> <span class="n">map_start</span>

<div class="viewcode-block" id="RecombinationMap.uniform_map"><a class="viewcode-back" href="../../utilities.html#msprime.RecombinationMap.uniform_map">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">uniform_map</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">rate</span><span class="p">,</span> <span class="n">num_loci</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a :class:`.RecombinationMap` instance in which the recombination</span>
<span class="sd">        rate is constant over a chromosome of the specified length.</span>
<span class="sd">        The legacy ``num_loci`` option is no longer supported and should not be used.</span>

<span class="sd">        :param float length: The length of the chromosome.</span>
<span class="sd">        :param float rate: The rate of recombination per unit of sequence length</span>
<span class="sd">            along this chromosome.</span>
<span class="sd">        :param int num_loci: This parameter is no longer supported.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">cls</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="n">length</span><span class="p">],</span> <span class="p">[</span><span class="n">rate</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">num_loci</span><span class="o">=</span><span class="n">num_loci</span><span class="p">)</span></div>

<div class="viewcode-block" id="RecombinationMap.read_hapmap"><a class="viewcode-back" href="../../utilities.html#msprime.RecombinationMap.read_hapmap">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">read_hapmap</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">filename</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parses the specified file in HapMap format.</span>

<span class="sd">        .. warning::</span>
<span class="sd">            This method is deprecated, use the module-level</span>
<span class="sd">            :func:`read_hapmap` function instead.</span>

<span class="sd">        :param str filename: The name of the file to be parsed. This may be</span>
<span class="sd">            in plain text or gzipped plain text.</span>
<span class="sd">        :return: A RecombinationMap object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="s2">&quot;RecombinationMap.read_hapmap() is deprecated. &quot;</span>
            <span class="s2">&quot;Use RateMap.read_hapmap() instead.&quot;</span><span class="p">,</span>
            <span class="ne">FutureWarning</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">rate_map</span> <span class="o">=</span> <span class="n">RateMap</span><span class="o">.</span><span class="n">read_hapmap</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">position_col</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">rate_col</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">rate</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">rate_map</span><span class="o">.</span><span class="n">rate</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">rate_map</span><span class="o">.</span><span class="n">position</span><span class="p">,</span> <span class="n">rate</span><span class="p">,</span> <span class="n">map_start</span><span class="o">=</span><span class="n">rate_map</span><span class="o">.</span><span class="n">start_position</span><span class="p">)</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">mean_recombination_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the weighted mean recombination rate</span>
<span class="sd">        across all windows of the entire recombination map.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">mean_rate</span>

<div class="viewcode-block" id="RecombinationMap.get_total_recombination_rate"><a class="viewcode-back" href="../../utilities.html#msprime.RecombinationMap.get_total_recombination_rate">[docs]</a>    <span class="k">def</span> <span class="nf">get_total_recombination_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the effective recombination rate for this genetic map.</span>
<span class="sd">        This is the weighted mean of the rates across all intervals.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">cumulative_mass</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span></div>

    <span class="k">def</span> <span class="nf">physical_to_genetic</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">get_cumulative_mass</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">genetic_to_physical</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genetic_x</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">cumulative_mass</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># If we have a zero recombination rate throughout then everything</span>
            <span class="c1"># except L maps to 0.</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_sequence_length</span><span class="p">()</span> <span class="k">if</span> <span class="n">genetic_x</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">genetic_x</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">cumulative_mass</span><span class="p">,</span> <span class="n">genetic_x</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="n">y</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">position</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
            <span class="o">+</span> <span class="p">(</span><span class="n">genetic_x</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">cumulative_mass</span><span class="p">[</span><span class="n">index</span><span class="p">])</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">rate</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">y</span>

    <span class="k">def</span> <span class="nf">physical_to_discrete_genetic</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">physical_x</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Discrete genetic space is no longer supported&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_per_locus_recombination_rate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Genetic loci are no longer supported&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_num_loci</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;num_loci is no longer supported&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_size</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_positions</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_rates</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">rate</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">get_sequence_length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">sequence_length</span>

    <span class="k">def</span> <span class="nf">get_length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># Deprecated: use get_sequence_length() instead</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_sequence_length</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">asdict</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">map</span><span class="o">.</span><span class="n">asdict</span><span class="p">()</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2015-2020, Tskit Developers

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>